File: wishlist-api/src/controller.js
import express from 'express';
import { requireAccessToken } from './auth0.js';
import Joi from '@hapi/joi';
import * as client from './client.js';

/**
 * Wishlist schema.
 */
const listSchema = Joi.object({
  name: Joi.string().required(),
}).options({ stripUnknown: true });

/**
 * Item schema.
 */
const itemSchema = Joi.object({
  description: Joi.string().required().empty(),
  url: Joi.string().optional().allow(null).uri(),
  price: Joi.string().optional().allow(null),
});

/**
 * Returns a list of wishlists owned by the authenticated user.
 *
 * @param {*} req
 * @param {*} res
 */
const getLists = (req, res) => {
  return client
    .getLists(req.profile.email)
    .then(data => {
      return res.json(data);
    })
    .catch(err => {
      console.error(`error retrieving a list of wishlists: ${err}`);
      return res.status(500).json({ error: err });
    });
};

/**
 * Returns a specified list if it exists.
 *
 * @param {*} _req
 * @param {*} res
 */
const getList = (req, res) => {
  const id = req.params.wishlistId;

  return client
    .getList(id)
    .then(data => {
      return res.json(data);
    })
    .catch(err => {
      console.error(`error retrieving wishlist with ID ${id}: ${err}`);
      return res.status(500).json({ error: err });
    });
};

/**
 * Creates a new list owned by the authenticated user.
 *
 * @param {*} req
 * @param {*} res
 */
const createList = (req, res) => {
  const result = listSchema.validate(req.body);

  if (result.error) {
    return res.status(400).json({ error: result.error.details.map(it => it.message) });
  }

  const name = result.value.name;
  const owner = req.profile.email;

  return client
    .createList(name, owner)
    .then(data => {
      return res.status(201).json(data);
    })
    .catch(err => {
      console.error(`error creating a wishlist: ${err}`);
      return res.status(500).json({ error: err });
    });
};

/**
 * Updates a list with new properties.
 *
 * @param {*} req
 * @param {*} res
 */
const updateList = (req, res) => {
  const id = req.params.wishlistId;
  const owner = req.profile.email;

  const result = listSchema.validate(req.body);

  if (result.error) {
    return res.status(400).json({ error: result.error.details.map(it => it.message) });
  }

  const name = result.value.name;

  return client
    .updateList(id, name, owner)
    .then(data => {
      return res.json(data);
    })
    .catch(err => {
      console.error(`error updating wishlist ${id}: ${err}`);
      return res.status(500).json({ error: err });
    });
};

/**
 * Deletes a list and any items associated with it.
 *
 * @param {*} req
 * @param {*} res
 */
const deleteList = (req, res) => {
  const id = req.params.wishlistId;
  const owner = req.profile.email;

  return client
    .deleteList(id, owner)
    .then(() => {
      return res.status(204).end();
    })
    .catch(err => {
      console.error(`error deleting wishlist ${id}: ${err}`);
      return res.status(500).json({ error: err });
    });
};

/**
 * Returns a list of items associated with the given wishlist ID.
 *
 * @param {*} req
 * @param {*} res
 */
const getItems = (req, res) => {
  const id = req.params.wishlistId;

  return client
    .getItems(id)
    .then(data => {
      return res.json(data);
    })
    .catch(err => {
      console.error(`error retrieving a items for wishlist ${id}: ${err}`);
      return res.status(500).json({ error: err });
    });
};

/**
 * Updates an item.
 *
 * @param {*} req
 * @param {*} res
 */
const updateItem = (req, res) => {
  const wishlistId = req.params.wishlistId;
  const itemId = req.params.itemId;
  const owner = req.profile.email;

  const result = itemSchema.validate(req.body);

  if (result.error) {
    return res.status(400).json({ error: result.error.details.map(it => it.message) });
  }

  const description = result.value.description;
  const url = result.value.url;
  const price = result.value.price;

  return client
    .updateItem(wishlistId, itemId, owner, description, url, price)
    .then(data => {
      return res.json(data);
    })
    .catch(err => {
      if (err instanceof client.AccessDeniedError) {
        res.status(403);
      } else {
        console.error(`error updating item ${itemId} in wishlist ${wishlistId}: ${err}`);
        res.status(500);
      }
      return res.json({ error: err });
    });
};

const deleteItem = (req, res) => {
  const wishlistId = req.params.wishlistId;
  const itemId = req.params.itemId;
  const owner = req.profile.email;

  client
    .deleteItem(wishlistId, itemId, owner)
    .then(() => {
      return res.status(204).end();
    })
    .catch(err => {
      if (err instanceof client.AccessDeniedError) {
        res.status(403);
      } else {
        console.error(`error deleting item ${itemId} in wishlist ${wishlistId}: ${err}`);
        res.status(500);
      }
      return res.json({ error: err });
    });
};

const createItem = (req, res) => {
  const wishlistId = req.params.wishlistId;

  const result = itemSchema.validate(req.body);

  if (result.error) {
    return res.status(400).json({ error: result.error.details.map(it => it.message) });
  }

  const description = result.value.description;
  const url = result.value.url;
  const price = result.value.price;

  const owner = req.profile.email;

  return client
    .createItem(wishlistId, owner, description, url, price)
    .then(data => {
      return res.status(201).json(data);
    })
    .catch(err => {
      if (err instanceof client.AccessDeniedError) {
        res.status(403);
      } else {
        console.error(`error adding an item to wishlist ${wishlistId}: ${err}`);
        res.status(500);
      }
      return res.json({ error: err });
    });
};

export const routes = () => {
  const router = new express.Router();

  router.use(express.json());
  router.get('/wishlists', requireAccessToken, getLists);
  router.post('/wishlists', requireAccessToken, createList);
  router.get('/wishlists/:wishlistId', getList);
  router.put('/wishlists/:wishlistId', requireAccessToken, updateList);
  router.delete('/wishlists/:wishlistId', requireAccessToken, deleteList);

  router.get('/wishlists/:wishlistId/items', getItems);
  router.put('/wishlists/:wishlistId/items/:itemId', requireAccessToken, updateItem);
  router.delete('/wishlists/:wishlistId/items/:itemId', requireAccessToken, deleteItem);
  router.post('/wishlists/:wishlistId/items', requireAccessToken, createItem);

  return router;
};

File: wishlist-api-python/src/controller.py
import json
from flask import Flask, jsonify, request, g
from flask_cors import CORS

from schema import ItemSchema, ListSchema
from marshmallow import ValidationError
from auth import requires_auth
from error import HttpException, AccessDeniedException
import client

app = Flask(__name__)
CORS(app)


@app.errorhandler(HttpException)
def handler_unauthenticated(e):
    response = jsonify({"error": e.error})
    response.status_code = e.status_code
    return response


@app.route("/", methods=["GET"])
def index():
    return jsonify({"status": "ok"})


@app.route("/wishlists", methods=["GET"])
@requires_auth
def get_lists():
    try:
        return jsonify(client.get_lists(g.profile["email"]))
    except Exception as e:
        print(f"error retrieving a list of wishlists: {str(e)}")
        return jsonify({"error": str(e)}), 500


@app.route("/wishlists", methods=["POST"])
@requires_auth
def create_list():
    try:
        parsed = ListSchema().load(request.json)
        result = client.create_list(g.profile["email"], parsed["name"])
        return jsonify(result), 201
    except ValidationError as e:
        return jsonify(e.messages), 400
    except Exception as e:
        print(f"error creating a wishlist: {str(e)}")
        return jsonify({"error": str(e)}), 500


@app.route("/wishlists/<id>", methods=["GET"])
def get_list(id):
    try:
        return jsonify(client.get_list(id))
    except Exception as e:
        print(f"error retrieving wishlist with ID {id}: {str(e)}")
        return jsonify({"error": str(e)}), 500


@app.route("/wishlists/<id>", methods=["PUT"])
@requires_auth
def update_list(id):
    try:
        parsed = ListSchema().load(request.json)
        result = client.update_list(id, parsed["name"], g.profile["email"])
        return jsonify(result)
    except ValidationError as e:
        return jsonify(e.messages), 400
    except Exception as e:
        print(f"error updating wishlist {id}: {e}")
        return jsonify({"error": str(e)}), 500


@app.route("/wishlists/<id>", methods=["DELETE"])
@requires_auth
def delete_list(id):
    try:
        client.delete_list(id, g.profile["email"])
        return "", 204
    except Exception as e:
        print(f"error deleting wishlist {id}: {str(e)}")
        return jsonify({"error": str(e)}), 500


@app.route("/wishlists/<list_id>/items", methods=["GET"])
def get_list_items(list_id):
    try:
        return jsonify(client.get_items(list_id))
    except Exception as e:
        print(f"error retrieving a items for wishlist {list_id}: {str(e)}")
        return jsonify({"error": str(e)}), 500


@app.route("/wishlists/<list_id>/items", methods=["POST"])
@requires_auth
def create_item(list_id):
    try:
        body = request.json

        if "url" in body and body["url"] == "":
            body["url"] = None

        parsed = ItemSchema().load(body)
        result = client.create_item(
            list_id,
            g.profile["email"],
            parsed["description"],
            parsed["url"],
            parsed["price"],
        )
        return jsonify(result), 201
    except ValidationError as e:
        return jsonify(e.messages), 400
    except AccessDeniedException as e:
        return jsonify({"error": str(e)}), 403
    except Exception as e:
        print(f"error adding an item to wishlist {list_id}: {e}")
        return jsonify({"error": str(e)}), 500


@app.route("/wishlists/<list_id>/items/<item_id>", methods=["PUT"])
@requires_auth
def update_item(list_id, item_id):
    try:
        body = request.json

        if "url" in body and body["url"] == "":
            body["url"] = None

        parsed = ItemSchema().load(body)
        result = client.update_item(
            list_id,
            item_id,
            g.profile["email"],
            parsed["description"],
            parsed["url"],
            parsed["price"],
        )
        return jsonify(result)
    except ValidationError as e:
        return jsonify(e.messages), 400
    except AccessDeniedException as e:
        return jsonify({"error": str(e)}), 403
    except Exception as e:
        print(f"error updating item {item_id} in wishlist {list_id}: {str(e)}")
        return jsonify({"error": str(e)}), 500


@app.route("/wishlists/<list_id>/items/<item_id>", methods=["DELETE"])
@requires_auth
def delete_item(list_id, item_id):
    try:
        client.delete_item(list_id, item_id, g.profile["email"])
        return "", 204
    except AccessDeniedException as e:
        return jsonify({"error": str(e)}), 403
    except Exception as e:
        print(f"error deleting item {item_id} in wishlist {list_id}: {str(e)}")
        return jsonify({"error": str(e)}), 500
